Mh 77  upload state requirements

sharrisoncaktus opened the pull request:
The actual file upload is not included, which will be done by MH-102.

sharrisoncaktus committed:
[mh-77] models: StateRequirement, StateRequirementLink

sharrisoncaktus committed:
Merge branch 'myhpom-develop' into mh-77--upload-state-requirements

* myhpom-develop:
  [mh-68] store healthcare_network element in a variable
  [MH-8] Adding .h6 to align baselines of floated items with their corresponding titles
  [MH-8] There shouldn't be a `hr` in base
  [MH-8] Small tweak to header structure (removing outer container to reduce embedding)
  [MH-8] Correctly pulling in data in template
  [MH-8] Making <hr> a bit more spacious on small screens
  [MH-8] Small responsiveness improvement
  [MH-8] Adding big asterisk
  [MH-8] Footer border styles
  [MH-8] Improved footer responsiveness
  [MH-8] Adding sidebar and footer
  [MH-8] Adding faq
  [MH-8] Start of basic dashboard layout
  [mh-68] bugfix
  Fixes migrations so that they are reversable (MH-60)
  Order territories at the bottom of 'ordered' states (MH-60)
  Adds territories to states model (MH-60).

sharrisoncaktus committed:
[mh-77] renamed migrations

sharrisoncaktus committed:
[mh-77] put migrations into sequence

sharrisoncaktus committed:
[mh-77] scaffold GET view and template w/o date picker

sharrisoncaktus committed:
[mh-77] integrate /upload/requirements into the app, rename template and provide basic .scss

sharrisoncaktus committed:
[mh-77] stub/docstring for testing StateRequirement model

sharrisoncaktus committed:
Merge branch 'myhpom-develop' into mh-77--upload-state-requirements

* myhpom-develop:
  [MH-91] Changing column sizes to accommodate breadcrumbs better
  [MH-91] Adding to the pattern library
  [MH-91] Adding some helpful comments
  [MH-91] Bumping up breadcrumb size on x-large screens
  [MH-91] Basic template, styles, and includes for upload widget

sharrisoncaktus committed:
[mh-77] integrate /upload/requirements with dashboard template

sharrisoncaktus committed:
[mh-77] /upload/requirements with styling and checkbox/button integration, but without date picker or file upload

sharrisoncaktus committed:
[mh-77] date picker for /upload/requirements, uses moment.js and combodate.js

sharrisoncaktus committed:
[mh-77] bold selected breadcrumb item

sharrisoncaktus committed:
[mh-77] style tweak

sharrisoncaktus committed:
[mh-77] incorporate combodate into enable/disable of SELECT FILE button.

sharrisoncaktus committed:
Merge commit 'ab6f3ae0e3651daeeadf1b6ffa604f99eef1d5da' into mh-77--upload-state-requirements

* commit 'ab6f3ae0e3651daeeadf1b6ffa604f99eef1d5da':
  To aid testing embed the upload templates in the dashboard (MH-27).
  Creates all upload endpoints (MH-27)
  Adds advanced directive model (MH-27)

# Conflicts:
#	myhpom/templates/myhpom/dashboard.html
#	myhpom/templates/myhpom/upload/requirements.html
#	myhpom/views/__init__.py

also merged migrations

sharrisoncaktus committed:
[mh-77]  dashboard now loads combodate.js in head for use by upload/requirements

sharrisoncaktus committed:
[mh-77] UploadRequirementsForm with one field corresponding to the AdvanceDirective.valid_date. validation requires a non-future date per discussion (though not per AC).

sharrisoncaktus committed:
[mh-77] upload/requirements now validates valid_date field both on client and server. (Client-side validation makes sense in this case to provide a better feedback loop.)

upload/submit no longer requires POST, because we're redirecting from upload/requirements via GET

sharrisoncaktus committed:
better validation error language

sharrisoncaktus committed:
Merge branch 'myhpom-develop' into mh-77--upload-state-requirements

* myhpom-develop:
  Clarify comment around inclusion of static serving
  [MH-69] Strip out most everything except...

sharrisoncaktus committed:
[mh-77] test StateRequirement, added pre-save validation to StateRequirement in order to provide more ValidationErrors

sharrisoncaktus committed:
[mh-77] test UploadRequirementForm

sharrisoncaktus committed:
[mh-77] tested StateRequirementLink

sharrisoncaktus committed:
[mh-77] test myhpom.views.directive.upload_requirements

sharrisoncaktus committed:
Merge branch 'myhpom-develop' into mh-77--upload-state-requirements

* myhpom-develop: (24 commits)
  [MH-101] The model changes weren't a good idea either, for that matter
  [MH-101] Relocating "submit" button & annotating forms properly
  [MH-101] That migration was never a good idea
  [MH-59] Adding to _blank link
  Adds pointers to links (MH-27)
  [MH-101] Similar custom network update logic for "no AD template" next steps view
  [MH-101] Moving over to using ModelForm; following "avoid null on CharField" recommendation
  [MH-59] MEDIA settings tweaks
  Removes unused template & improves button styling (MH-27)
  Adds tests for sharing step (MH-27)
  Adds styling for active breadcrumbs (MH-27)
  Add styleguide elements for upload widget (MH-27)
  Adds styling for sharing part of upload widget (MH-27)
  Makes consistent button styling for widget (MH-27)
  Fleshes out the signup markup (MH-27)
  Fixes test case to account for guard statement (MH-27)
  Styles widget index page (MH-27)
  Adds sharing checkbox form, and styles 'current status' (MH-27)
  Adds initial tests for upload views (MH-27)
  Fixes dashboard tests (MH-27)
  ...

# Conflicts:
#	myhpom/forms/upload_requirements.py
#	myhpom/static/myhpom/include/_upload_widget.scss
#	myhpom/templates/myhpom/upload/requirements.html
#	myhpom/views/directive.py

sharrisoncaktus committed:
[mh-77] undo

sharrisoncaktus committed:
[mh-77] resolved merge conflicts, but not done reconfiguring

sharrisoncaktus committed:
[mh-77] add validator to AdvanceDirective.valid_date – must be in the past when saved. (This was previously on the UploadRequirementsForm.)

sharrisoncaktus committed:
[mh-77] updated upload_requirements view to fit current structure (moved from views.directive to views.upload, among other small changes)

sharrisoncaktus committed:
[mh-77] upload_requirements template updates post merge

dsummersl commented:
Can one of the admins verify this patch?

dsummersl commented:
Tiny request - it appears that Neil left out an import for `User` here https://github.com/myhpom/MyHPOM/blob/9d602b5f9cdede97bfd8afabaab9fad2805afa1c/myhpom/tests/test_next_steps_view.py#L44 (and its the only broken test on your branch). Mind throwing in an import statement on your branch?

dsummersl commented on myhpom/static/myhpom/page/_upload_requirements.scss:
```diff
@@ -0,0 +1,55 @@
+
+$invalid_color: #dc3545;
+
+.upload-requirements__div {
```
This doesn't appear to be used anywhere; could be removed.

sharrisoncaktus replied:
ah, yes, I restyled to match what you had done. 

I'll be glad to fix that little testing bug – just wasn't sure if I should.

dsummersl commented on myhpom/static/myhpom/page/_upload_requirements.scss:
```diff
@@ -0,0 +1,55 @@
+	margin: 0 0 1rem 0;
+}
+
+.upload-requirements__ol {
```
I think the names in this scss file should be revisited in a few cases. From my reading of BEM the `element` should be the conceptual element that documents the purpose of the CSS rather than describe the structural element. The elements that are most opaque for me are those with elements of `__ol`, `__h4`, `__li`, `__a` -- names like `__list`, `__subtitle`, `__list__item`, `__learn-more` will help devs understand the intent of the class a little more quickly.

dsummersl commented on myhpom/models/state_requirement_link.py:
```diff
@@ -0,0 +1,31 @@
+from myhpom.validators import validate_not_blank
+
+
+class StateRequirementLink(models.Model):
```
Does a link need its own model? These could be `default=''` fields on the StateRequirement and make for a simpler model.

sharrisoncaktus replied:
It's because some of the StateRequirements have multiple links, so it's not one-to-one but one-to-many

dsummersl replied:
Oof, so it is! I generally hate the idea of stuffing html into the database, so IMO putting an arbitrary html field into SRs isn't a great idea. So: scratch my suggestion.

dsummersl commented on myhpom/models/state_requirement_link.py:
```diff
@@ -0,0 +1,31 @@
+    """
+
+    requirement = models.ForeignKey(StateRequirement, on_delete=models.CASCADE)
+    text = models.CharField(max_length=1024)
```
Would we ever expect to display 1k characters - I think a much smaller max makes sense here.

sharrisoncaktus replied:
256 then

dsummersl commented on myhpom/models/state_requirement.py:
```diff
@@ -0,0 +1,52 @@
+    * UNIQUE(state, text) -- protect against multiple items with the same text
+    """
+
+    state = models.ForeignKey(State, on_delete=models.CASCADE, null=True, blank=True)
```
For us to test this feature on staging I think this model will need an ModelAdmin associated with it. In the admin it would be helpful to have `help_text` on each field.

dsummersl replied:
hindsight...an inline admin form on the existing states admin would be the most directly useful to a tester, since these can't exist without a state anyway.

sharrisoncaktus replied:
Not entirely sure what you're envisioning for the admin, so how about let's defer that / make a separate ticket. But I did add the help_text.

dsummersl commented on myhpom/templates/myhpom/upload/requirements.html:
```diff
@@ -13,12 +13,134 @@
 {% block widget_body %}
+<div class="row">
+	<div class="col">
+		<h3 class="h6">ADVANCE DIRECTIVE DOCUMENT REQUIREMENTS FOR {{ user.userdetails.state.title|upper }}</h3>
```
Neil has poked me a couple times to let the CSS to the uppercasing for content like this, so I'm poking you ;) - Add a class like `upload-requirements__title` and use `text-transform: uppercase;`.

dsummersl commented on myhpom/templates/myhpom/upload/requirements.html:
```diff
@@ -13,12 +13,134 @@
+		<form id="id_requirements_form" class="upload-requirements__form" method="POST" action="{% url 'myhpom:upload_requirements' %}">
+			{% csrf_token %}
+			<div id="id_valid_date_div" class="upload-requirements__input-div{% if form.errors.valid_date %}--invalid{% endif %}">
+				<input id="id_valid_date" name="valid_date" data-min-year="{{ MIN_YEAR }}" data-max-year="{{ MAX_YEAR }}"{% if form.errors.valid_date %} class="upload-requirement__input--invalid"{% endif %}/>
```
To make the styling consistent for this third party library attach the appropriate bootstrap classes to the classes that are produced...like

```
.combodate > .month {
  @extends .form-control;
}
```

(I *think* that would work - but I haven't tried it so YMMV!)

sharrisoncaktus replied:
yep, that works, but with a small tweak

dsummersl commented on myhpom/templates/myhpom/upload/requirements.html:
```diff
@@ -13,12 +13,134 @@
+    var $inputSelectFile = $("#id_input_select_file");
+
+	function submitButtonEnableDisable () {
+		console.log('file = ' + $inputSelectFile.val());
```
Remove debug logging.

dsummersl commented on myhpom/templates/myhpom/upload/requirements.html:
```diff
@@ -13,12 +13,134 @@
+		console.log('file = ' + $inputSelectFile.val());
+		if ($('.upload-requirements__checkbox:checked').length == $requirements.length
+			&& $directiveDateInput.combodate('getValue') != ''
+			&& $inputSelectFile.val().trim() != ''
```
Convert these to `===` and `!==`.

dsummersl commented on myhpom/templates/myhpom/upload/requirements.html:
```diff
@@ -13,12 +13,134 @@
+			&& $directiveDateInput.combodate('getValue') != ''
+			&& $inputSelectFile.val().trim() != ''
+		) {
+    		$submitButton.prop('disabled', null).css('cursor', 'pointer');
```
Rather than setting properties on the button create two classes and add/remove the class as appropriate (say `.upload-requirements__submit` and `.upload-requirements__submit--disabled`). This keeps styling/behavior  cleanly separated.

sharrisoncaktus replied:
except that the disabled prop can't be fully emulated in css...? I want the pointer event to fire (and change the cursor) but not to allow the click. Otherwise, I can just add "pointer-events: none;" to the --disabled class.

dsummersl replied:
True, I've no issue with the disabled property. I'm concerned with setting css properties in jquery. Rather than setting individual css properties, use classes.

dsummersl commented on myhpom/templates/myhpom/upload/requirements.html:
```diff
@@ -13,12 +13,134 @@
 {% endblock widget_breadcrumb_items %}
```
The additions to this file are a mix of space and tab delineated. Use space delineated (4 spaces per tab stop).

sharrisoncaktus replied:
good to know, thanks. and now I've updated my Sublime settings to make this automatic.

dsummersl replied:
+1 I'm sorry we don't have some linter integrated to catch little things like this :/

dsummersl reviewed (changes requested):
Still working my way thru this PR, but here are some suggestions I have thus far. Manually testing it out, behaves well!

dsummersl commented on myhpom/static/myhpom/page/_upload_requirements.scss:
```diff
@@ -0,0 +1,55 @@
+/*
```
This file also uses tabs - use spaces (4 spaces to a tab stop)

dsummersl commented on myhpom/templates/myhpom/upload/requirements.html:
```diff
@@ -13,12 +13,134 @@
+
+    $requirements.on('change', function (event) {
+    	submitButtonEnableDisable();
+    });
```
This can be simplified to `$requirements.on('change', submitButtonEnableDisable);`

dsummersl commented on myhpom/templates/myhpom/upload/requirements.html:
```diff
@@ -13,12 +13,134 @@
+
+	$inputSelectFile.on('change', function (event) {
+		submitButtonEnableDisable();
+	});
```
This can also be simplified as per above.

dsummersl commented on myhpom/tests/test_directive_upload_requirements_view.py:
```diff
@@ -0,0 +1,44 @@
+        response = self.client.post(self.url, data=form_data)
+        self.assertRedirects(
+            response, reverse('myhpom:upload_sharing'), fetch_redirect_response=False
+        )
```
Assert that the date has been saved to the model?

dsummersl commented on myhpom/tests/test_directive_upload_requirements_view.py:
```diff
@@ -0,0 +1,44 @@
+        self.client.login(username=user_data['email'], password=user_data['password'])
+        self.url = reverse('myhpom:upload_requirements')
+        state = models.State.objects.get(name="NC")
+        userdetails = models.UserDetails.objects.create(user=self.user, state=state)
```
You can simplify setup by using the `UserFactory` (see myhpom/tests/test_next_steps_view.py for examples).

dsummersl commented on myhpom/tests/test_state_requirement_model.py:
```diff
@@ -0,0 +1,77 @@
+            #     'state': None,
+            #     'position': len(self.global_reqs),
+            #     'text': self.global_reqs[0].text,  # violate uniqueness for state, text
+            # },
```
Does this commented out block need to be another test case?

sharrisoncaktus replied:
Yes, I just needed to spend a little more time and raise the IntegrityError in the model.

dsummersl commented on myhpom/views/upload.py:
```diff
@@ -35,6 +28,37 @@ def upload_current_ad(request):
+            else:
+                directive = AdvanceDirective(user=request.user, share_with_ehs=False)
+            directive.valid_date = form.cleaned_data['valid_date']
+            directive.save()
```
Should not need to explicitly set `valid_date` here, the ModelForm should do that during `save()`.

sharrisoncaktus committed:
[mh-77] test_next_steps_view small bugfix

sharrisoncaktus committed:
[mh-77] fix styles per BEM

sharrisoncaktus committed:
[mh-77] js cleanup, also disable input submit if directiveDate > NOW

sharrisoncaktus committed:
[mh-77] "Let CSS do the uppercasing," he poked. ;-)

sharrisoncaktus committed:
[mh-77] much shorter text for StateRequirementLink.text

sharrisoncaktus committed:
[mh-77] "To make the styling consistent for this third party library attach the appropriate bootstrap classes to the classes that are produced."

sharrisoncaktus committed:
[mh-77] now we need !important to overcome bootstrap

sharrisoncaktus committed:
[mh-77] typed comparison

sharrisoncaktus committed:
[mh-77] classes for upload and upload--disabled; fix whitespace

sharrisoncaktus committed:
[mh-77] simplify, reduce, declutter

sharrisoncaktus committed:
[mh-77] "Assert that the date has been saved to the model"

sharrisoncaktus committed:
[mh-77] "You can simplify setup by using the UserFactory"

sharrisoncaktus committed:
[mh-77]
dsummersl: "Does this commented out block need to be another test case?"
sharrisoncaktus: "Yes, I just needed to spend a little more time and raise the IntegrityError in the model."

sharrisoncaktus committed:
[mh-77] "Should not need to explicitly set valid_date here, the ModelForm should do that during save()."

sharrisoncaktus committed:
[mh-77] "In the admin it would be helpful to have help_text on each field."

dsummersl commented on myhpom/templates/myhpom/upload/requirements.html:
```diff
@@ -14,57 +14,57 @@
+        <form id="id_requirements_form" class="upload-requirements__form" method="POST" action="{% url 'myhpom:upload_requirements' %}">
+            {% csrf_token %}
+            <div id="id_valid_date_div">
+                <input id="id_valid_date" name="valid_date" data-min-year="{{ MIN_YEAR }}" data-max-year="{{ MAX_YEAR }}"/>
```
This field is crammed together with the above/below. Wrapping it in a `p` (or some other bs4 block) would make it match the other forms margins.

dsummersl commented on myhpom/static/myhpom/combodate-1.0.7/README.md:
`````diff
@@ -0,0 +1,25 @@
+
+## Installation
+````
+bower install combodate
`````
Ideally we don't have 3rd party libraries checked into the repository...seeing that this wasn't in npm and can't be included there I thought yesterday, "I guess we don't have any choice." But...then I saw this bower install bit. We could/should use it to install rather than checking this in. I'd suggest adding bower as a dev dependency (I can do this cleanup bit in my ticket if you'd rather punt - I realize hacking docker & setting up bower & static includes isn't the funnest).

sharrisoncaktus replied:
It looks like adding the following to package.json dependencies will do the trick:
    "combodate": "vitalets/combodate#1.1.0"
and no need for bower.

dsummersl replied:
awesome!

dsummersl commented on myhpom/static/myhpom/page/_upload_requirements.scss:
```diff
@@ -0,0 +1,72 @@
+.combodate > .month,
+.combodate > .day {
+  @extend .form-control;
+  display: inline-block;
```
4 spaces per indent. (otherwise this worked and the change looks much better)

dsummersl commented on myhpom/tests/test_next_steps_view.py:
```diff
@@ -3,6 +3,7 @@
 from django.core.files.uploadedfile import SimpleUploadedFile
 
 from myhpom.models import State
+from myhpom.models.user import User
```
ty!

dsummersl reviewed (changes requested):
couple small things (and a question around the 3rd party lib). Getting very close!

sharrisoncaktus committed:
[mh-77] "This field is crammed together with the above/below." The <ol> that the requirements are in could use a little top/bottom margin, which also solves this issue.

sharrisoncaktus committed:
[mh-77] "Ideally we don't have 3rd party libraries checked into the repository." It looks like adding the following to package.json dependencies will do the trick: "combodate": "vitalets/combodate#1.1.0" -- and no need for bower.

dsummersl commented on myhpom/migrations/0010_state_requirements.py:
```diff
@@ -0,0 +1,41 @@
+
+    operations = [
+        migrations.CreateModel(
+            name='StateRequirement',
```
This migration appears to need a refresh - when I try this locally I need to run makemigrations again to capture the help_text changes.

sharrisoncaktus replied:
Ah, ok, sorry about that.

sharrisoncaktus replied:
And also the addition of a validator to the AdvanceDirective model

sharrisoncaktus committed:
[mh-77] "4 spaces per indent."

sharrisoncaktus committed:
Merge branch 'myhpom-develop' into mh-77--upload-state-requirements

* myhpom-develop:
  [mh-77] The 2-story-point version of the task is better than the 3-story-point version.
  Revert "[mh-11] /download/template displays the AD template for the user's state, else 404"
  Revert "[mh-11] add AD template download, if available, to dashboard upload/index widget."
  Revert "[mh-11] moved dashboard tests to own file, added test for presence / absence of advance_template_directive based on user's state (which determines whether the "Download a Template" button should be shown)"
  Revert "[mh-11] tests for /download/template"
  [mh-11] tests for /download/template
  [mh-11] moved dashboard tests to own file, added test for presence / absence of advance_template_directive based on user's state (which determines whether the "Download a Template" button should be shown)
  [mh-11] add AD template download, if available, to dashboard upload/index widget.
  [mh-11] /download/template displays the AD template for the user's state, else 404

sharrisoncaktus committed:
[mh-77] "This migration appears to need a refresh - when I try this locally I need to run makemigrations again to capture the help_text changes." And also the addition of a validator to the AdvanceDirective model

dsummersl reviewed (approved):
🎆 we're there!
